src/databricks/labs/ucx/migration/azure_credentials.py

```diff
@@ -0,0 +1,170 @@
+                   client_id=storage_permission_mapping.client_id,
+                   principal=storage_permission_mapping.principal,
+                   privilege=storage_permission_mapping.privilege,
+                   client_secret=client_secret)
+
+
+class AzureServicePrincipalMigration:
+
+    def __init__(self, installation: Installation, ws: WorkspaceClient, azure_resource_permissions: AzureResourcePermissions,
+                 azure_sp_crawler: AzureServicePrincipalCrawler):
+        self._final_sp_list = None
+        self._installation = installation
+        self._ws = ws
+        self._azure_resource_permissions = azure_resource_permissions
+        self._azure_sp_crawler = azure_sp_crawler
+        self._action_plan = 'service_principals_for_storage_credentials.csv'
+
+
+    @classmethod
+    def for_cli(cls, ws: WorkspaceClient, customized_csv: str, replace_with_ac: bool, product='ucx'):
+        installation = Installation.current(ws, product)
+        config = installation.load(WorkspaceConfig)
+        sql_backend = StatementExecutionBackend(ws, config.warehouse_id)
+        azurerm = AzureResources(ws)
+        locations = ExternalLocations(ws, sql_backend, config.inventory_database)
+
+        azure_resource_permissions = AzureResourcePermissions(installation, ws, azurerm, locations)
+        azure_sp_crawler = AzureServicePrincipalCrawler(ws, sql_backend, config.inventory_database)
+
+        return cls(installation, ws, azure_resource_permissions, azure_sp_crawler, customized_csv, replace_with_ac)
+
+
+    def _list_storage_credentials(self):
+        # list existed storage credentials that is using service principal, capture the service principal's application_id
+        return {}
+
+
+    def _check_sp_in_storage_credentials(self, sp_list, sc_set):
+        # if sp is already used, take it off from the sp_list
+        return list()
+
+
+    def _fetch_client_secret(self, sp_list: list[StoragePermissionMapping]) -> list[ServicePrincipalMigrationInfo]:
+        # check AzureServicePrincipalInfo from AzureServicePrincipalCrawler, if AzureServicePrincipalInfo
+        # has secret_scope and secret_key not empty, fetch the client_secret and put it to the client_secret field
+        #
+        # The input StoragePermissionMapping may have managed identity mixed in, we will ignore them for now, as
+        # they won't have any client_secret, we will process managed identity in the future.
+
+        # fetch client_secrets of crawled service principal, if any
+        azure_sp_info_with_client_secret = {}
+        azure_sp_infos = self._azure_sp_crawler.snapshot()
+
+        for azure_sp_info in azure_sp_infos:
+            if azure_sp_info.secret_scope is None:
+                continue
+            if azure_sp_info.secret_key is None:
+                continue
+
+            try:
+                secret_response = self._ws.secrets.get_secret(azure_sp_info.secret_scope, azure_sp_info.secret_key)
```

Comment:
Are you 1000% sure this call won't fail outside DBR? Did you test it on labs-azure? **When I implemented dbutils in the sdk, it wasn't the case.**

Create a new task/workflow in runtime.py and call this method of migration from there (within a job).

Then inject WorkspaceInstallation into this class and ".run_workflow()" it from here.

------

@nkvuong this is the context for you why you need to fix the crawler.

Reply:
this is correct - You create secrets using the REST API or CLI, but you must use the [Secrets utility (dbutils.secrets)](https://docs.databricks.com/en/dev-tools/databricks-utils.html#dbutils-secrets) in a notebook or job to read a secret.

Reply:
Yes, I have tested fetching secret from field-eng-east workspace in my local laptop with `az login`:
<img width="634" alt="image" src="https://github.com/databrickslabs/ucx/assets/91635877/fa760aae-4d7c-430e-bbf6-e8f950a7f904">

Reply:
@qziyuan That workspace probably has a custom setting enabled. Try it on labs-azure workspace.

Reply:
@nfx , just tried the labs-azure workspace, same behavior:
<img width="1369" alt="image" src="https://github.com/databrickslabs/ucx/assets/91635877/f32a4aa5-51cd-43f8-a3fc-1ec27785349b">